### PR TITLE
Add metrics tracking for eviction to ResponseCache

### PR DESCRIPTION
**Pull request by realtyem**
It appears to be a slight oversight that tracking time based removal was not added. Here you go.

response_cache_eviction_results

I think it looks pretty good. I used the query:

```
rate(synapse_util_caches_response_cache_evicted_size{instance=~"$instance", job=~"$job", index=~"$index"}[$bucket_size])
```

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Jason Little <realtyem@gmail.com>

### CODE REVIEW

File: synapse/util/caches/response_cache.py [hunk 1]
```diff
@@ -187,6 +187,16 @@ def unset(self, key: KV) -> None:
...
         self._result_cache.pop(key, None)
 
+    def evict_because(self, key: KV, reason: EvictionReason) -> None:
+        """Basically the same as unset, but update reason why evicting for metrics"""
+        self._metrics.inc_evictions(reason)
+        self.unset(key)
+
+    def entry_timeout(self, key: KV) -> None:
+        """For the call_later to remove from the cache"""
+        logger.debug(f"Expiring from [{self._name}] - {key}")
```
Comment (realtyem) [inline synapse/util/caches/response_cache.py:197]: I am open to removing the debugging log lines, but it's up to you.

**Comment by clokep** [comment]
@realtyem You mentioned to me that we do this for "regular" caches; can you link to that code? It should help with review.

**Comment by realtyem** [comment]
> @realtyem You mentioned to me that we do this for "regular" caches; can you link to that code? It should help with review.

By "regular" caches, I meant like `LruCache` which uses a completely different set of mechanisms for evicting 'entries'. Consider: jemalloc, background loops, cache-invalidation-over-replication, etc. The `ResponseCache` is by nature a much, much simpler implementation. I can try to pin it down, which one would you like to see?

File: synapse/util/caches/response_cache.py [hunk 2]
```diff
@@ -187,6 +187,16 @@ def unset(self, key: KV) -> None:
         """
         self._result_cache.pop(key, None)
 
+    def evict_because(self, key: KV, reason: EvictionReason) -> None:
+        """Basically the same as unset, but update reason why evicting for metrics"""
+        self._metrics.inc_evictions(reason)
+        self.unset(key)
+
+    def entry_timeout(self, key: KV) -> None:
```
Comment (clokep) [inline synapse/util/caches/response_cache.py:195]: This seems reasonable, but I have a few suggestions:

* Let's make this private (add a `_` in front of it).
* Get rid of the `evict_because` method -- it is trivial and doesn't add much IMO / only adds indirection.
* Stop calling `unset` and inline it.
* Update `unset` to include `self._metrics.inc_evictions(EvicationReason.invalidation)`.

Comment (realtyem) [reply synapse/util/caches/response_cache.py]: I think I got it all into [59563dd](https://github.com/matrix-org/synapse/pull/16028/commits/59563ddddad7f1bb4c59cf321e4334e496f20236)

**Review by clokep — APPROVED** [review]

**Comment by clokep** [comment]
Thanks!

**Pull request merged**